core/src/main/java/org/apache/calcite/config/CalciteSystemProperty.java

```diff
@@ -176,6 +176,14 @@
   public static final CalciteSystemProperty<Boolean> TEST_SLOW =
       booleanProperty("calcite.test.slow", false);
 
+  /**
+   * Whether to do validation within VolcanoPlanner after each rule firing.
+   * Note that doing so would significantly slow down the planning. Should only
+   * enable for unit test.
+   */
+  public static final CalciteSystemProperty<Boolean> TEST_VALIDATE_VOLCANO_PLANNER =
+      booleanProperty("calcite.test.validate.volcano.planner", false);
```

Comment:
I think the flag is not limited to test, so we can remove `test` from the flag name.
TEST_VALIDATE_VOLCANO_PLANNER -> VALIDATE_VOLCANO_PLANNER

Reply:
Doesn't seem appropriate as a system property. Maybe a tracing option.

Reply:
Do you mean log trace level? I update this because I don't want to tight it with log level. I hope in the future we can turn this check on by default in unit test. It doesn't necessarily enable tracing at the same time.

Reply:
I don't want to add too many system properties that are meaningless to end-users.

You can add a tracer that doesn't print stuff, just controls validation.

It's dangerous to turn extra validation on in tests. The validation can cause side-effects that cause code to work in tests, fail in production.

Another idea is to do the extra validation if asserts are enabled (`-ea`). Run the tests with asserts enabled and disabled.

Reply:
The validation shouldn't have any side-effects. If there are any, I think we should just fix them. IMO the extra validation is quite useful as we have seen different consistency issues with the memo. 

-ea seems to be a good idea. Will look into that.

Reply:
Maybe not in this case, but validation CAN have side effects such as ensuring that a cache or memo is populated (e.g. `AbstractRelNode.deriveRowType()`); in production, the cache is not populated, so the code fails.

Reply:
Regarding `-ea`. This is what we are using when we do `assert isValid(Litmus.THROW);`.

Reply:
+1 on `assert`.